### PR TITLE
tickets/DM-9818

### PR DESCRIPTION
**Pull request by athornton**
This implements only a class that provides for a data fetch and render.

I'm going to build the driver that creates a complete HTML page on top of this, but I want to be able to push the class up to PyPI in order to have the module available for the service/Kubernetes job when I do a pip install inside a container.

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1,6 +1,7 @@
-MIT License
+The MIT License (MIT)
```
Comment (jonathansick) [inline LICENSE]: Don't make this change.

File: .gitignore [hunk 2]
```diff
@@ -43,47 +43,16 @@ htmlcov/
...
-.python-version
-
-# celery beat schedule file
-celerybeat-schedule
-
-# dotenv
-.env
-
-# virtualenv
-venv/
```
Comment (jonathansick) [inline .gitignore:82]: Why did you need to change the default `.gitignore` provided by GitHub for Python projects?

File: LICENSE (hunk 1 above)
Comment (jonathansick) [inline LICENSE]: Just make sure that the GitHub API still registers this license as the MIT License after you change stuff like this (the license text is seeded by GitHub)

File: dochubproto/dochubproto.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        if mca:
+            try:
+                self.max_document_data_age = float(mca)
+            except ValueError:
+                self.warning("Could not convert %s '%s' to number" %
+                             (mdda, mca))
+        self.document_refresh_time = 0.0  # The epoch stands in for "never"
+        self.state = "empty"
+
+    def debug(self, *args, **kwargs):
```
Comment (jonathansick) [inline dochubproto/dochubproto.py:101]: Unusual, but okay.

File: dochubproto/dochubproto.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+        """
+        if self.logger:
+            self.logger.critical(*args, **kwargs)
+
+    def start_document_data_refresh(self):
+        """Start document data refresh in a separate thread, and return
+        immediately.  Progress can be monitored by polling with the
+        check_status() method.
+        """
+        thd = Thread(target=self.refresh_document_data)
```
Comment (jonathansick) [inline dochubproto/dochubproto.py:136]: I'm hesitant to be using threads at this stage, but okay.

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python
+"""Setuptools script.
+"""
+import os
+import codecs
+from setuptools import setup, find_packages
+
+PACKAGENAME = 'sqre-dochub-proto'
+DESCRIPTION = 'Bootstrapper for cookiecutter projects'
```
Comment (jonathansick) [inline setup.py:9]: Update this

File: dochubproto/dochubproto.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+                self.state = self.STATE_STALE
+            finally:
+                mutex.release()
+        return self.state
+
+    def render(self, refresh_if_stale=False):
+        """Render document data into HTML.  If refresh_if_stale is True,
+        check document freshness, and refresh if necessary before rendering.
+        """
+        sdd = self.get_document_data()
```
Comment (jonathansick) [inline dochubproto/dochubproto.py:238]: Might want to improve this variable name so I know what this is.

**Review by jonathansick — APPROVED** [review]
I'm a bit surprised by the use of threading here in this prototype, but that's fine. I'd try to separate data scraping and HTML rendering in separate classes, but that's fine too at this stage. So far so good.

**Comment by jonathansick** [comment]
@athornton You need to create a separate ticket for this work, you can't highjack my ticket which is already marked "Done".

**Review by jonathansick — CHANGES_REQUESTED** [review]
Create a new ticket branch before merging.

File: dochubproto/dochubproto.py (hunk 4 above)
Comment (athornton) [reply dochubproto/dochubproto.py:136]: Makes it immensely faster.  The whole render only takes a few seconds.
